ledger/byzcoin/mod.go

```diff
@@ -0,0 +1,249 @@
+	}
+
+	go ldgr.routine(bcActor, players)
+
+	return newActor(ldgr.gossiper), err
+}
+
+func (ldgr *Ledger) routine(actor blockchain.Actor, players mino.Players) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	blocks := ldgr.bc.Watch(ctx)
+
+	roundTimeout := time.After(initialRoundTime)
+
+	for {
+		select {
+		// TODO: closing
+		case rumor := <-ldgr.gossiper.Rumors():
+			tx, ok := rumor.(Transaction)
+			if ok {
+				ldgr.queue.Add(tx)
+			}
+		case <-roundTimeout:
+			// This timeout has two purposes. The very first use will determine
+			// the round time before the first block is proposed after a boot.
+			// Then it will be used to insure that blocks are still proposed in
+			// of catastrophic failure in the consensus layer (i.e. too many
+			// players offline for a while).
+			go ldgr.proposeBlock(actor, players)
+
+			roundTimeout = time.After(timeoutRoundTime)
+		case block := <-blocks:
+			payload := block.GetPayload().(*BlockPayload)
+
+			txRes := make([]TransactionResult, len(payload.GetTxs()))
+			for i, pb := range payload.GetTxs() {
+				tx, err := ldgr.txFactory.FromProto(pb)
+				if err != nil {
+					return
+				}
+
+				txRes[i] = TransactionResult{
+					txID: tx.(Transaction).hash,
+				}
+			}
+
+			ldgr.queue.Remove(txRes...)
+
+			// This is executed in a different go routine so that the gathering
+			// of transactions can keep on while the block is created.
+			go ldgr.proposeBlock(actor, players)
+
+			roundTimeout = time.After(timeoutRoundTime)
+		}
+	}
+}
+
+func (ldgr *Ledger) proposeBlock(actor blockchain.Actor, players mino.Players) {
+	payload, err := ldgr.makePayload(ldgr.queue.GetAll())
+	if err != nil {
+		fabric.Logger.Err(err).Msg("couldn't make the payload")
+	}
+
+	// Each instance proposes a payload based on the received
+	// transactions but it depends on the blockchain implementation
+	// if it will be accepted.
+	err = actor.Store(payload, players)
+	if err != nil {
+		fabric.Logger.Err(err).Msg("couldn't send the payload")
+	}
+}
+
+func (ldgr *Ledger) makePayload(txs []Transaction) (*BlockPayload, error) {
+	payload := &BlockPayload{
+		Txs: make([]*TransactionProto, 0, len(txs)),
+	}
+
+	for _, tx := range txs {
+		packed, err := tx.Pack()
+		if err != nil {
+			return nil, err
+		}
+
+		payload.Txs = append(payload.Txs, packed.(*TransactionProto))
+	}
+
+	return payload, nil
+}
+
+// Watch implements ledger.Ledger. It listens for new transactions and returns
+// the transaction result that can be used to verify if the transaction has been
+// accepted or rejected.
+func (ldgr *Ledger) Watch(ctx context.Context) <-chan ledger.TransactionResult {
+	blocks := ldgr.bc.Watch(ctx)
+	results := make(chan ledger.TransactionResult)
+
+	go func() {
+		for {
+			block, ok := <-blocks
+			if !ok {
+				fabric.Logger.Debug().Msg("watcher is closing")
+				return
+			}
+
+			payload := block.GetPayload().(*BlockPayload)
+
+			for _, pb := range payload.GetTxs() {
```

Comment:
if my understanding is correct, I would find it easier to follow if `pb` were renamed to `tx` and `tx` to `txProto`

Reply:
Only `pb` to `txProto` but you're right.